lib/api/bucketPut.js

```diff
@@ -39,36 +56,77 @@ export default function bucketPut(authInfo, request, locationConstraint, log,
         log.trace('invalid acl header');
         return callback(errors.InvalidArgument);
     }
-    // - AWS JS SDK sends a request with locationConstraint us-east-1
-    // if no locationConstraint provided.
-    if (locationConstraint && config.locationConstraints &&
-      Object.keys(config.locationConstraints).indexOf(locationConstraint) < 0) {
-        log.trace('locationConstraint is invalid',
-          { locationConstraint });
-        return callback(errors.InvalidLocationConstraint);
-    }
-
-    if (!locationConstraint && config.restEndpoints && request.parsedHost &&
-      config.restEndpoints[request.parsedHost]) {
-        locationConstraintChecked = config.restEndpoints[request.parsedHost];
-    } else {
-        locationConstraintChecked = locationConstraint;
-    }
-    const bucketName = request.bucketName;
+    const { bucketName } = request;
 
-    return createBucket(authInfo, bucketName, request.headers,
-        locationConstraintChecked, config.usEastBehavior, log,
-        (err, previousBucket) => {
-            // if bucket already existed, gather any relevant cors headers
-            const corsHeaders = collectCorsHeaders(request.headers.origin,
-                request.method, previousBucket);
-            if (err) {
-                return callback(err, corsHeaders);
+    return waterfall([
+        next => _parseXML(request, log, next),
+        // Check policies in Vault for a user.
+        (locationConstraint, next) => {
+            if (authInfo.isRequesterAnIAMUser()) {
+                const authParams = auth.server.extractParams(request, log, 's3',
+                    request.query);
+                const requestContextParams = {
+                    constantParams: {
+                        headers: request.headers,
+                        query: request.query,
+                        generalResource: bucketName,
+                        specificResource: '',
+                        requesterIp: request.socket.remoteAddress,
+                        sslEnabled: request.connection.encrypted,
+                        apiMethod: 'bucketPut',
+                        awsService: 's3',
+                        locationConstraint,
+                        requesterInfo: authInfo,
+                        signatureVersion: authParams.params.data.authType,
+                        authType: authParams.params.data.signatureVersion,
+                        signatureAge: authParams.params.data.signatureAge,
+                    },
+                };
+                return vault.checkPolicies(requestContextParams,
+                    authInfo.getArn(), log, (err, authorizationResults) => {
+                        if (err) {
+                            return next(err);
+                        }
+                        if (authorizationResults[0].isAllowed === false) {
```

Comment:
1) just to be careful, let's reverse this and do if authorizationResults[0].isAllowed !== true
2) let's add a log saying that the authorization check failed